lib/client/local_proxy_middleware.go

```diff
@@ -29,103 +30,194 @@ import (
 		log.WithError(err).Warn("unable to load profile, requesting database certs without access requests")
 	} else {
 		accessRequests = profile.ActiveRequests.AccessRequests
 	}
 
 	var key *Key
-	if err := RetryWithRelogin(ctx, c.tc, func() error {
-		newKey, err := c.tc.IssueUserCertsWithMFA(ctx, ReissueParams{
-			RouteToCluster: c.tc.SiteName,
+	if err := RetryWithRelogin(ctx, c.Client, func() error {
+		newKey, err := c.Client.IssueUserCertsWithMFA(ctx, ReissueParams{
+			RouteToCluster: c.Client.SiteName,
 			RouteToDatabase: proto.RouteToDatabase{
-				ServiceName: c.dbRoute.ServiceName,
-				Protocol:    c.dbRoute.Protocol,
-				Username:    c.dbRoute.Username,
-				Database:    c.dbRoute.Database,
+				ServiceName: c.RouteToApp.ServiceName,
+				Protocol:    c.RouteToApp.Protocol,
+				Username:    c.RouteToApp.Username,
+				Database:    c.RouteToApp.Database,
 			},
 			AccessRequests: accessRequests,
 			RequesterName:  proto.UserCertsRequest_TSH_DB_LOCAL_PROXY_TUNNEL,
-		}, mfa.WithPromptReasonSessionMFA("database", c.dbRoute.ServiceName))
+		}, mfa.WithPromptReasonSessionMFA("database", c.RouteToApp.ServiceName))
 		key = newKey
 		return trace.Wrap(err)
 	}); err != nil {
-		return trace.Wrap(err)
+		return tls.Certificate{}, trace.Wrap(err)
 	}
 
-	dbCert, ok := key.DBTLSCerts[c.dbRoute.ServiceName]
-	if !ok {
-		return trace.NotFound("database '%v' TLS cert missing", c.dbRoute.ServiceName)
-	}
-	tlsCert, err := keys.X509KeyPair(dbCert, key.PrivateKeyPEM())
+	dbCert, err := key.DBTLSCert(c.RouteToApp.ServiceName)
 	if err != nil {
+		return tls.Certificate{}, trace.Wrap(err)
+	}
+	return dbCert, nil
+}
+
+// AppCertIssuer checks and issues app certs.
+type AppCertIssuer struct {
+	// Client is a TeleportClient used to issue certificates when necessary.
+	Client *TeleportClient
+	// RouteToApp contains app routing information.
+	RouteToApp proto.RouteToApp
+}
+
+func (c *AppCertIssuer) CheckCert(cert *x509.Certificate) error {
+	// appCertIssuer does not perform any additional certificate checks.
+	return nil
+}
+
+func (c *AppCertIssuer) IssueCert(ctx context.Context) (tls.Certificate, error) {
+	var accessRequests []string
+	if profile, err := c.Client.ProfileStatus(); err != nil {
+		log.WithError(err).Warn("unable to load profile, requesting database certs without access requests")
+	} else {
+		accessRequests = profile.ActiveRequests.AccessRequests
+	}
+
+	var key *Key
+	if err := RetryWithRelogin(ctx, c.Client, func() error {
+		appCertParams := ReissueParams{
+			RouteToCluster: c.Client.SiteName,
+			RouteToApp:     c.RouteToApp,
+			AccessRequests: accessRequests,
+			RequesterName:  proto.UserCertsRequest_TSH_APP_LOCAL_PROXY,
+		}
+
+		// TODO (Joerger): DELETE IN v17.0.0
+		clusterClient, err := c.Client.ConnectToCluster(ctx)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		rootClient, err := clusterClient.ConnectToRootCluster(ctx)
+		if err != nil {
+			return trace.Wrap(err)
+		}
```

Comment:
does this dial the root cluster? do you think it would be possible/worth avoiding it, if clusterClient is the root cluster, or the session isn't necessary? could be a followup PR, i know this is already used elsewhere

Reply:
If `clusterClient` is root, it just returns the auth client already dialed:

```go
// ConnectToRootCluster connects to the auth server of the root cluster
// via proxy. It returns connected and authenticated auth server client.
func (c *ClusterClient) ConnectToRootCluster(ctx context.Context) (auth.ClientI, error) {
	root, err := c.ConnectToCluster(ctx, c.root)
	return root, trace.Wrap(err)
}

// ConnectToCluster connects to the auth server of the given cluster via proxy. It returns connected and authenticated auth server client
func (c *ClusterClient) ConnectToCluster(ctx context.Context, clusterName string) (auth.ClientI, error) {
	if c.cluster == clusterName {
		return c.CurrentCluster(), nil
	}
...
```